Telemetry: Removing duplicate config in http tracing

jakub-pomykala opened the pull request:
- Update the logic that checks `"otel.trace.http.disabled"` to avoid extracting all of the otel config into a map (which is complex and expensive)
  - it's not necessary here, we can just query MP Config directly
- Adding in a test to check the correct amount of spans is returned when `"otel.trace.http.disabled"` is `true` 
- Adding in a test where servlet tracing is disabled and we call a JAX-RS resource

Fixes #26804


Also:
- Removing a left over `system.out.println()` in `JaxRsRouteTestServlet` class

jakub-pomykala committed:
Removing duplicate config in http tracing

- Adding in test for when OTEL is disabled
- Removing debug sys.out from JaxRsRouteTestServlet class (Left over from previous work)

Yas-Franco commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/TelemetryDisabledServletTest.java:
```diff
@@ -0,0 +1,87 @@
+import io.opentelemetry.sdk.autoconfigure.spi.traces.ConfigurableSpanExporterProvider;
+
+/**
+ * HTTP request tracing tests
```
This comment needs updating

jakub-pomykala replied:
Done

Yas-Franco commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/servlet/DisabledServlet.java:
```diff
@@ -0,0 +1,79 @@
+    public static final String CONTEXT_ASYNC_SERVLET = "contextAsync";
+    public static final String PLACEHOLDER_SERVLET = "placeholder";
+    public static final String HELLO_HTML = "hello.html";
+    public static final String DICE_JSP = "dice.jsp";
```
You don't need most of these constants

jakub-pomykala replied:
Done

Yas-Franco commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/TelemetryDisabledServletTest.java:
```diff
@@ -0,0 +1,87 @@
+                        .addPackage(AbstractSpanMatcher.class.getPackage())
+                        .addAsServiceProvider(ConfigurableSpanExporterProvider.class, InMemorySpanExporterProvider.class)
+                        .addAsWebResource(new File("test-applications/TelemetryServletTestApp.war/hello.html"))
+                        .addAsWebResource(new File("test-applications/TelemetryServletTestApp.war/dice.jsp"))
```
I don't think you use those web resources in this test so you can probably remove them

jakub-pomykala replied:
Done

jakub-pomykala committed:
Removing un-needed constants from DisabledServerlet test

- Updating comment previously copied from different test

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.1.internal/src/io/openliberty/microprofile/telemetry11/internal/rest/TelemetryServletFilter.java:
```diff
@@ -312,17 +309,6 @@ public List<String> getHttpResponseHeader(final ServletRequest request, final Se
 
     }
 
-    private HashMap<String, String> getTelemetryProperties() {
-        HashMap<String, String> telemetryProperties = new HashMap<>();
-        for (String propertyName : config.getPropertyNames()) {
-            if (propertyName.startsWith("otel") || propertyName.startsWith("OTEL")) {
-                config.getOptionalValue(propertyName.toLowerCase().replace('_', '.'), String.class).ifPresent(
-                    value -> telemetryProperties.put(propertyName.toLowerCase().replace('_', '.'), value));
-            }
-        }
-        return telemetryProperties;
-    }
-
```
The method below this one can be removed as well I think.

Azquelt replied:
I also think these changes apply to the other version of `TelemetryServletFilter` that we have.

jakub-pomykala replied:
Done

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/servlet/DisabledServlet.java:
```diff
@@ -0,0 +1,74 @@
+    @Inject
+    private HttpServletRequest request;
+
+    @Test
+    public void testSimpleServlet() throws Exception {
+        String scheme = request.getScheme();
+        String serverName = request.getServerName();
+        int serverPort = request.getServerPort();
+        URL url = new URL(scheme + "://" + serverName + ":" + serverPort + "/" + APP_NAME + "/" + SIMPLE_SERVLET);
+
+        String traceId = httpGet(url); // The servlet outputs the traceId
+
+        // The simple servlet will return the default invalid traceId
+        assertEquals(traceId, INVALID_TRACE_ID);
+
+    }
```
This test is valid and I think we should keep it, but it doesn't really test what you're changing here.

I think the only way to actually find out if the change has broken something would be to enable trace and look for the "instrumenter is initialized" or "instrumenter is set to null" messages.

We could add that check to this test, but I think it's sufficient to check it yourself manually before merging this PR.

jakub-pomykala replied:
Run with tracing enabled which did return "instrumenter is set to null"

Azquelt reviewed (changes requested):
These changes look mostly good. It would be helpful to add a justification in the PR description explaining why removing the call to `checkDisabled()` is correct and where the duplicated code is located.

jakub-pomykala committed:
Remove the unused checkDisabled()

- Adding the same changes to telemetry-1.0
- Removing unused checkDisabled() method

jakub-pomykala commented:
#build
#spawn.fullfat.buckets=io.openliberty.microprofile.telemetry.1.0.internal_fat,io.openliberty.microprofile.telemetry.1.0.internal.jaeger_fat,io.openliberty.microprofile.telemetry.1.0.internal_fat_tck,io.openliberty.microprofile.telemetry.1.1.internal_fat_tck

jakub-pomykala commented:
#build
#spawn.fullfat.buckets=io.openliberty.microprofile.telemetry.1.0.internal_fat,io.openliberty.microprofile.telemetry.1.0.internal.jaeger_fat,io.openliberty.microprofile.telemetry.1.0.internal_fat_tck,io.openliberty.microprofile.telemetry.1.1.internal_fat_tck

jakub-pomykala committed:
Adding in http tracing check

- Fixing broken personal builds in telemetry 1.0 and 1.1
- Removed unused import from TElemetryDisabledServletTest class
- Add in new Test for checking HttpTrace to test if it is disabled

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry10/internal/rest/TelemetryServletFilter.java:
```diff
@@ -101,14 +99,16 @@ private Instrumenter<ServletRequest, ServletResponse> getInstrumenter() {
     }
 
     private Instrumenter<ServletRequest, ServletResponse> createInstrumenter() {
+        // Check if the HTTP tracing should be disabled
+        boolean httpTracingDisabled = config.getOptionalValue("otel.trace.http.disabled", Boolean.class).orElse(false);
         OpenTelemetryInfo otelInfo = OpenTelemetryAccessor.getOpenTelemetryInfo();
         if (tc.isDebugEnabled()) {
             Tr.debug(tc, "otelInfo.getEnabled()=" + otelInfo.getEnabled());
         }
```
Please trace the value of `httpTracingDisabled`

jakub-pomykala replied:
Done

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/TelemetryDisabledServletTest.java:
```diff
@@ -0,0 +1,83 @@
+import io.openliberty.microprofile.telemetry.internal_fat.shared.spans.AbstractSpanMatcher;
+import io.opentelemetry.sdk.autoconfigure.spi.traces.ConfigurableSpanExporterProvider;
+
+/**
+ * Test with opentelemetry disabled
+ */
+@RunWith(FATRunner.class)
+public class TelemetryDisabledServletTest extends FATServletClient {
```
We already have a `TelemetryDisabledTest`. It looks like this test is also disabling otel with `otel.sdk.disabled`.

If there's something new to test, should it be added to the existing test instead?

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/servlet/DisabledServlet.java:
```diff
@@ -0,0 +1,74 @@
+
+import componenttest.app.FATServlet;
+
+@SuppressWarnings("serial")
+@WebServlet("/testDisabledHttpTrace")
+public class DisabledServlet extends FATServlet {
```
Please rename the class to `HttpTraceDisabledServlet` to better describe what it's testing.

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/servlet/DisabledServlet.java:
```diff
@@ -0,0 +1,74 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package io.openliberty.microprofile.telemetry.internal_fat.apps.servlet;
```
This servlet does not run in the same app as the rest of the classes in this package.

If you need a new app to test this (and I think you do, because you're testing different configuration), put it in a new package.

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/TelemetryHttpTraceDisabledTest.java:
```diff
@@ -0,0 +1,84 @@
+/**
+ * Test with otel.trace.http.disabled set to true
+ */
+@RunWith(FATRunner.class)
+public class TelemetryHttpTraceDisabledTest extends FATServletClient {
+
+    public static final String APP_NAME = "TelemetryServletTestApp";
+    public static final String SERVER_NAME = "Telemetry10Servlet";
+
+    @TestServlets({
+                    @TestServlet(contextRoot = APP_NAME, servlet = DisabledServlet.class),
+    })
+
+    @Server(SERVER_NAME)
+    public static LibertyServer server;
```
I don't think you need a new test class for this test. It could probably be a second app in `TelemetryServletTest`. This will avoid starting the server an extra time and make the test run more quickly.

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.1.internal/src/io/openliberty/microprofile/telemetry11/internal/rest/TelemetryServletFilter.java:
```diff
@@ -96,16 +93,18 @@ private Instrumenter<ServletRequest, ServletResponse> getInstrumenter() {
-    
+
     private Instrumenter<ServletRequest, ServletResponse> createInstrumenter() {
+        // Check if the HTTP tracing should be disabled
+        boolean httpTracingDisabled = config.getOptionalValue("otel.trace.http.disabled", Boolean.class).orElse(false);
         OpenTelemetryInfo otelInfo = OpenTelemetryAccessor.getOpenTelemetryInfo();
         if (tc.isDebugEnabled()) {
             Tr.debug(tc, "otelInfo.getEnabled()=" + otelInfo.getEnabled());
         }
```
Please trace out the value of `httpTracingDisabled` here.

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/servlet/DisabledServlet.java:
```diff
@@ -0,0 +1,74 @@
+    @Inject
+    private HttpServletRequest request;
+
+    @Test
+    public void testSimpleServlet() throws Exception {
+        String scheme = request.getScheme();
+        String serverName = request.getServerName();
+        int serverPort = request.getServerPort();
+        URL url = new URL(scheme + "://" + serverName + ":" + serverPort + "/" + APP_NAME + "/" + SIMPLE_SERVLET);
+
+        String traceId = httpGet(url); // The servlet outputs the traceId
+
+        // The simple servlet will return the default invalid traceId
+        assertEquals(traceId, INVALID_TRACE_ID);
+
+    }
```
This test is fine, but we should also add another test where servlet tracing is disabled and we call a JAX-RS resource (so we still expect spans from JAX-RS)

jakub-pomykala replied:
Done

jakub-pomykala committed:
Adding in new httpTraceDisabledTest

- Adding in new test for when http tracing is disabled
- Moving old test to work on TelemetryServletTest server instead of a new one
- Tracing value for httpTracingDisabled
- Removing TelemetryDisabledServletTest as a test for this already exists
- Renaming DisabledServlet class to HttpTraceDisabledServlet and moving to new package

jakub-pomykala committed:
Merge remote-tracking branch 'upstream/integration' into duplicate-config-tracing

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry10/internal/rest/TelemetryServletFilter.java:
```diff
@@ -101,14 +99,17 @@ private Instrumenter<ServletRequest, ServletResponse> getInstrumenter() {
 
     private Instrumenter<ServletRequest, ServletResponse> createInstrumenter() {
+        // Check if the HTTP tracing should be disabled
+        boolean httpTracingDisabled = config.getOptionalValue("otel.trace.http.disabled", Boolean.class).orElse(false);
```
Should use the `CONFIG_DISABLE_HTTP_TRACING_PROPERTY` constant that the removed code used for the config key.

jakub-pomykala committed:
Changing the httpTracingDisabled check

- changing to look for CONFIG_DISABLE_HTTP_TRACING_PROPERTY instead of otel.trace.http.disabled

jakub-pomykala commented:
#build
#spawn.fullfat.buckets=io.openliberty.microprofile.telemetry.1.0.internal_fat,io.openliberty.microprofile.telemetry.1.0.internal.jaeger_fat,io.openliberty.microprofile.telemetry.1.0.internal_fat_tck,io.openliberty.microprofile.telemetry.1.1.internal_fat_tck

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/publish/servers/Telemetry10Servlet/server.xml:
```diff
@@ -9,11 +9,14 @@
         <classloader apiTypeVisibility="+third-party" />
     </application>
     
+<application type="war" location="HttpTraceDisabledServletTestApp.war">
+		<classloader apiTypeVisibility="+third-party" />
+	</application>
```
Try to match the indentation in the rest of the file.

Azquelt reviewed (dismissed):
There's one comment above but it's not important enough to prevent merging.

Azquelt commented:
#libby

jakub-pomykala committed:
Fixing check for httpTracingDisabled

- Using CONFIG_DISABLE_HTTP_TRACING_PROPERTY constant
- Changing testSimpleServlet as application is now only disabling http tracing and not otel.sdk therefore testSimpleServlet will not work anymore and now testing staticFile when http tracing is disabled
- Fixing formatting in server.xml to match

jakub-pomykala committed:
Work in progress: Fix for testDisabledStaticFile()

- testDisabledStaticFile still not functioning as expected

jakub-pomykala committed:
testing JSP when http tracing is disabled

- changing testDisabledStaticFile to testDisabledJsp (still the same idea just that the hello.html file was wont working as expected and was only returning 404
-

jakub-pomykala committed:
Merge remote-tracking branch 'upstream/integration' into duplicate-config-tracing

jakub-pomykala commented:
#build
#spawn.fullfat.buckets=io.openliberty.microprofile.telemetry.1.0.internal_fat,io.openliberty.microprofile.telemetry.1.0.internal.jaeger_fat,io.openliberty.microprofile.telemetry.1.0.internal_fat_tck,io.openliberty.microprofile.telemetry.1.1.internal_fat_tck

jakub-pomykala commented:
#libby

jakub-pomykala committed:
Add testSimpleServlet

- When http.trace is disabled testSimpleServlet will check the traceId and make sure that it is invalid as expected

Azquelt commented on dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/publish/servers/Telemetry10Servlet/server.xml:
```diff
@@ -9,11 +9,14 @@
         <classloader apiTypeVisibility="+third-party" />
     </application>
     
+    <application type="war" location="HttpTraceDisabledServletTestApp.war">
+		<classloader apiTypeVisibility="+third-party" />
+    </application>
```
Indentation is inconsistent here.

jakub-pomykala committed:
Fixing indentation issue

jakub-pomykala commented:
#build
#spawn.fullfat.buckets=io.openliberty.microprofile.telemetry.1.0.internal_fat,io.openliberty.microprofile.telemetry.1.0.internal.jaeger_fat,io.openliberty.microprofile.telemetry.1.0.internal_fat_tck,io.openliberty.microprofile.telemetry.1.1.internal_fat_tck

jakub-pomykala commented:
#libby

jakub-pomykala commented:
The build failed due to unrelated defects.

Everything required for Telemetry passed.
